src/ralph/api/routers/statements.py

```diff
@@ -0,0 +1,244 @@
+    ),
+    ###
+    # Private use query string parameters
+    ###
+    search_after: Optional[str] = Query(
+        None,
+        description=(
+            "Sorting data to allow pagination through large number of search results. "
+            "NB: for internal use, not part of the LRS specification."
+        ),
+    ),
+    pit_id: Optional[str] = Query(
+        None,
+        description=(
+            "Point-in-time ID to ensure consistency of search requests through "
+            "multiple pages."
+            "NB: for internal use, not part of the LRS specification."
+        ),
+    ),
+):
+    """Fetch a single xAPI Statement or multiple xAPI Statements."""
+
+    # Create a list to aggregate all the enabled filters for the search
+    es_query_filters = []
+
+    if statementId:
+        es_query_filters += [{"term": {"_id": statementId}}]
+
+    if agent:
+        es_query_filters += [{"term": {"actor.account.name.keyword": agent}}]
+
+    if verb:
+        es_query_filters += [{"term": {"verb.id.keyword": verb}}]
+
+    if activity:
+        es_query_filters += [
+            {"term": {"object.objectType.keyword": "Activity"}},
+            {"term": {"object.id.keyword": activity}},
+        ]
+
+    if since:
+        es_query_filters += [{"range": {"timestamp": {"gt": since}}}]
+
+    if until:
+        es_query_filters += [{"range": {"timestamp": {"lte": until}}}]
+
+    if len(es_query_filters) > 0:
+        es_query = {"query": {"bool": {"filter": es_query_filters}}}
+    else:
+        es_query = {"query": {"match_all": {}}}
+
+    # Honor the "ascending" parameter, otherwise show most recent statements first
+    es_query.update(
+        {"sort": [{"timestamp": {"order": "asc" if ascending else "desc"}}]}
+    )
+
+    # Create a point-in-time or use the existing one to ensure consistency of search
+    # results over multiple pages.
+    if not pit_id:
+        # pylint: disable=unexpected-keyword-arg
+        pit_id = ES_CLIENT.open_point_in_time(index="statements", keep_alive="1m")["id"]
+    es_query.update(
+        {
+            "pit": {
+                "id": pit_id,
+                "keep_alive": "1m",  # extend duration of PIT whenever it is used
+            }
+        }
+    )
+
+    if search_after:
+        es_query.update({"search_after": search_after.split("|")})
+
+    # Make sure the limit does not go above max from settings
+    limit = min(limit, ES_MAX_SEARCH_HITS_COUNT)
+
+    # pylint: disable=unexpected-keyword-arg
+    es_response = ES_CLIENT.search(
+        body=es_query,
+        size=limit,
+    )
+
+    # Prepare the link to get the next page of requests, while preserving the
+    # consistency of search results.
+    # NB: There is an unhandled edge case where the total number of results is
+    # exactly a multiple of the "limit", in which case we'll offer an extra page
+    # with 0 results.
+    response = {}
+    if len(es_response["hits"]["hits"]) == limit:
```

Comment:
The number of hits is in the ES response right?

```suggestion
    if es_response["hits"]["total"]["value"] == limit:
```

Reply:
So `es_response["hits"]["total"]["value"]` is not very useful in my opinion: it (pretends to) represents the total number of hits, not the number of hits contained in the current response.

I say `pretends to` because it is capped at an arbitrary number defined by settings (which is 10000 by default and is not recommended to be increased to a large number).


If we want to better manage pagination, IMO, the way is to use a custom aggregation that simply counts the total number of matching documents. I did not pursue this as the current implementation (only having a minor issue in an edge case) seemed good enough to me in the context of this POC.